File: ApolloCore/src/main/java/me/solar/apollo/apolloCore/utils/MathsUtils.java
package me.solar.apollo.apolloCore.utils;

import java.util.*;

public final class MathsUtils {

    private MathsUtils() {}

    public static int add(int a, int b) { return a + b; }
    public static int subtract(int a, int b) { return a - b; }
    public static int multiply(int a, int b) { return a * b; }
    public static int divide(int a, int b) { return a / b; }

    public static Map<Integer, String> getRomanNumerals() {
        Map<Integer, String> map = new LinkedHashMap<>();
        map.put(1, "I"); map.put(2, "II"); map.put(3, "III"); map.put(4, "IV"); map.put(5, "V");
        map.put(6, "VI"); map.put(7, "VII"); map.put(8, "VIII"); map.put(9, "IX"); map.put(10, "X");
        map.put(40, "XL"); map.put(50, "L"); map.put(90, "XC"); map.put(100, "C"); map.put(400, "CD");
        map.put(500, "D"); map.put(900, "CM"); map.put(1000, "M");
        return map;
    }

    public static String convertToRomanNumeral(int number) {
        Map<Integer, String> romanNumerals = getRomanNumerals();
        int remainingNumber = number;
        StringBuilder result = new StringBuilder();
        List<Integer> keys = new ArrayList<>(romanNumerals.keySet());
        Collections.sort(keys, Collections.reverseOrder());
        for (int key : keys) {
            while (remainingNumber >= key) {
                result.append(romanNumerals.get(key));
                remainingNumber -= key;
            }
        }
        return result.toString();
    }

    public static int max(int[] a) {
        int max = a[0];
        for (int i = 1; i < a.length; i++) if (a[i] > max) max = a[i];
        return max;
    }

    public static int min(int[] a) {
        int min = a[0];
        for (int i = 1; i < a.length; i++) if (a[i] < min) min = a[i];
        return min;
    }

    public static double floor(double a) { return Math.floor(a); }
    public static double ceil(double a) { return Math.ceil(a); }
    public static long round(double a) { return Math.round(a); }
    public static double abs(double a) { return Math.abs(a); }
    public static boolean inRange(int a, int b, int c) { return a >= b && a <= c; }
    public static boolean inRange(double a, double b, double c) { return a >= b && a <= c; }
    public static double increasePercentage(double a, double b) { return a + percentage(a, b); }
    public static double decreasePercentage(double a, double b) { return a - percentage(a, b); }
    public static double percentage(double a, double b) { return a * b / 100; }
    public static double percentageChange(double a, double b) { return ((b - a) / a) * 100; }
    public static double average(double[] a) {
        double sum = 0.0;
        for (double v : a) sum += v;
        return sum / a.length;
    }
    public static double median(double[] a) {
        double[] copy = Arrays.copyOf(a, a.length);
        Arrays.sort(copy);
        if (copy.length % 2 == 0)
            return (copy[copy.length / 2] + copy[copy.length / 2 - 1]) / 2;
        else
            return copy[copy.length / 2];
    }
    public static double mode(double[] a) {
        Map<Double, Integer> freq = new HashMap<>();
        for (double v : a) freq.put(v, freq.getOrDefault(v, 0) + 1);
        double mode = a[0];
        int maxCount = 0;
        for (Map.Entry<Double, Integer> entry : freq.entrySet()) {
            if (entry.getValue() > maxCount) {
                maxCount = entry.getValue();
                mode = entry.getKey();
            }
        }
        return mode;
    }
    public static int factorial(int a) {
        int result = 1;
        for (int i = 1; i <= a; i++) result *= i;
        return result;
    }
    public static int fibonacci(int a) {
        if (a <= 1) return a;
        return fibonacci(a - 1) + fibonacci(a - 2);
    }
    public static boolean isPrime(int a) {
        if (a <= 1) return false;
        for (int i = 2; i <= a / 2; i++) if (a % i == 0) return false;
        return true;
    }
    public static boolean isEven(int a) { return a % 2 == 0; }
    public static boolean isOdd(int a) { return a % 2 != 0; }
    public static boolean isPositive(int a) { return a > 0; }
    public static boolean isNegative(int a) { return a < 0; }
    public static boolean isZero(int a) { return a == 0; }
    public static boolean isDivisibleBy(int a, int b) { return a % b == 0; }
    public static boolean isPerfectSquare(int a) { return Math.sqrt(a) % 1 == 0.0; }
    public static boolean isPerfectCube(int a) { return Math.cbrt(a) % 1 == 0.0; }
    public static boolean isPowerOfTwo(int a) { return a != 0 && (a & (a - 1)) == 0; }
}

File: ApolloCore/src/main/kotlin/me/solar/apollo/apolloCore/utils/MathsUtils.kt
@file:JvmName("MathsUtils")
package me.solar.apollo.apolloCore.utils

import kotlin.math.abs
import kotlin.math.cbrt
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.round
import kotlin.math.sqrt

fun add(a: Int, b: Int): Int {
    return a + b
}

fun subtract(a: Int, b: Int): Int {
    return a - b
}

fun multiply(a: Int, b: Int): Int {
    return a * b
}

fun divide(a: Int, b: Int): Int {
    return a / b
}

fun getRomanNumerals(): Map<Int, String> {
    return mapOf(
        1 to "I",
        2 to "II",
        3 to "III",
        4 to "IV",
        5 to "V",
        6 to "VI",
        7 to "VII",
        8 to "VIII",
        9 to "IX",
        10 to "X",
        40 to "XL",
        50 to "L",
        90 to "XC",
        100 to "C",
        400 to "CD",
        500 to "D",
        900 to "CM",
        1000 to "M"
    )
}

fun convertToRomanNumeral(number: Int): String {
    val romanNumerals = getRomanNumerals()
    var remainingNumber = number
    var result = ""
    while (remainingNumber > 0) {
        val highestValue = romanNumerals.keys.last { it <= remainingNumber }
        result += romanNumerals[highestValue]
        remainingNumber -= highestValue
    }
    return result
}

fun max(a: Array<Int>) : Int {
    var max = a[0]
    for (i in 1 until a.size) {
        if (a[i] > max) {
            max = a[i]
        }
    }
    return max;
}

fun min(a: Array<Int>) : Int {
    var min = a[0]
    for (i in 1 until a.size) {
        if (a[i] < min) {
            min = a[i]
        }
    }
    return min;
}

fun floor(a: Double) : Double {
    return floor(a);
}

fun ceil(a: Double) : Double {
    return ceil(a);
}

fun round(a: Double) : Long {
    return round(a).toLong();
}

fun abs(a: Double) : Double {
    return abs(a);
}

fun inRange(a: Int, b: Int, c: Int) : Boolean {
    return a in b..c;
}

fun inRange(a: Double, b: Double, c: Double) : Boolean {
    return a in b..c;
}

fun increasePercentage(a: Double, b: Double) : Double {
    return a + percentage(a, b);
}

fun decreasePercentage(a: Double, b: Double) : Double {
    return a - percentage(a, b);
}

fun percentage(a: Double, b: Double) : Double {
    return (a * b / 100);
}

fun percentageChange(a: Double, b: Double) : Double {
    return ((b - a) / a) * 100;
}

fun average(a: Array<Double>) : Double {
    var sum = 0.0;
    for (i in 0 until a.size) {
        sum += a[i];
    }
    return sum / a.size;
}

fun median(a: Array<Double>) : Double {
    a.sort();
    if (a.size % 2 == 0) {
        return (a[a.size / 2] + a[a.size / 2 - 1]) / 2;
    } else {
        return a[a.size / 2];
    }
}

fun mode(a: Array<Double>) : Double {
    var mode = a[0];
    var maxCount = 0;
    for (i in 0 until a.size) {
        var count = 0;
        for (j in 0 until a.size) {
            if (a[j] == a[i]) {
                count++;
            }
        }
        if (count > maxCount) {
            maxCount = count;
            mode = a[i];
        }
    }
    return mode;
}

fun factorial(a: Int) : Int {
    var result = 1;
    for (i in 1..a) {
        result *= i;
    }
    return result;
}

fun fibonacci(a: Int) : Int {
    if (a <= 1) {
        return a;
    }
    return fibonacci(a - 1) + fibonacci(a - 2);
}

fun isPrime(a: Int) : Boolean {
    if (a <= 1) {
        return false;
    }
    for (i in 2..a / 2) {
        if (a % i == 0) {
            return false;
        }
    }
    return true;
}

fun isEven(a: Int) : Boolean {
    return a % 2 == 0;
}

fun isOdd(a: Int) : Boolean {
    return a % 2 != 0;
}

fun isPositive(a: Int) : Boolean {
    return a > 0;
}

fun isNegative(a: Int) : Boolean {
    return a < 0;
}

fun isZero(a: Int) : Boolean {
    return a == 0;
}

fun isDivisibleBy(a: Int, b: Int) : Boolean {
    return a % b == 0;
}

fun isPerfectSquare(a: Int) : Boolean {
    return sqrt(a.toDouble()) % 1 == 0.0;
}

fun isPerfectCube(a: Int) : Boolean {
    return cbrt(a.toDouble()) % 1 == 0.0;
}

fun isPowerOfTwo(a: Int) : Boolean {
    return a != 0 && (a and a - 1) == 0;
}


